Actions - Add nonMRI workflow - allow failure - JRuby, TruffleRuby [changelog skip]

MSP-Greg opened the pull request:
### Description

Add JRuby & TruffleRuby to Actions via another 'job', all are 'allow-failure'

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [ ] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

MSP-Greg commented:
At present, Actions does not have an 'allow-failures' similar to Travis/AppVeyor.  Last I checked, one must still use the web UI to actually look at the steps to see whether a failure occurred, as the failure 'indication' won't propagate any further.

So, given the CI here and in my fork, one option is to have a non-MRI workflow that allows failures and runs JRuby and TruffleRuby on Unbuntu.  A separate workflow makes it easier to force a seed value for testing, limit tests, etc.  Once the CI stabilizes, then move JRuby & TruffleRuby back into the main CI.

JFYI, setup-ruby-pkgs uses the code from my fork of setup-ruby, and the selection/installation of Ruby builds is identical to setup-ruby.  If you open the step, you'll see the same output during the 'install' phase...

nateberkopec commented:
@MSP-Greg But a separate workflow would still have the same issue of requiring us to view the output to see if it failed or not, right?

MSP-Greg commented:
@nateberkopec

Yes.  The main reason for a separate workflow (could be in the same file) is it would not need logic to determine which jobs failed.

nateberkopec commented:
Makes sense. Let's do it. Also for now, I think we should be on truffleruby-head because they're fixing a few issues re: frozen strings that our build is surfacing.

MSP-Greg commented:
> truffleruby-head

Will do, almost done.

eregon commented:
Agreed, for now there is no point to test with `truffleruby`, just test with `truffleruby-head`.

MSP-Greg commented:
Any thoughts on Jruby, ie `jruby-9.2` or `jruby-head`?

nateberkopec commented on .github/workflows/ruby.yml:
```diff
@@ -40,10 +40,10 @@ jobs:
         run:  |
           # RubyGems < 3, no frozen strings
-          if [[ "2.6" < "${{ matrix.ruby }}.0" ]]; then
+          if [[ "2.6" < "${{ matrix.ruby }}.0" ]] && [[ "${{ matrix.ruby }}" != "truffleruby" ]]; then
```
truffleruby not needed here, right?

MSP-Greg replied:
Removed here and line 146, as truffleruby-head is ok

nateberkopec commented on .github/workflows/ruby.yml:
```diff
@@ -85,10 +85,56 @@ jobs:
+      fail-fast: false
+      matrix:
+        cfg:
+          - { os: ubuntu-latest, ruby: jruby-9.2        }
```
JRuby 9.2 failure should not be allowed - that build passes and is one of our supported platforms.

nateberkopec replied:
Personally I'd just put it in the main action and call this one nonMRIHead or something

MSP-Greg replied:
Done

eregon commented on .github/workflows/ruby.yml:
```diff
@@ -85,10 +85,55 @@ jobs:
+        if: success()
+        run:  |
+          # RubyGems < 3, no frozen strings
+          if [[ "2.6" < "${{ matrix.cfg.ruby }}.0" ]]; then
```
This check won't work, since there is no version in the matrix entries above

MSP-Greg replied:
Please check the log of the 'load ruby, ragel' step, and you'll see the correct matrix.cfg.ruby version loaded from setup-ruby's log.  This is another supported way to construct a matrix.

eregon replied:
It results in checks like:
```
# RubyGems < 3, no frozen strings
  if [[ "2.6" < "jruby-head.0" ]]; then
    export RUBYOPT="--enable=frozen-string-literal"
  fi
```
which don't make much sense (https://github.com/puma/puma/pull/2185/checks?check_run_id=509502033).

MSP-Greg replied:
It was my understanding that jruby-head supported `--enable=frozen-string-literal`.  Is that correct?  As I thought the conditional is true...

nateberkopec replied:
I don't really care as long it evals correctly FWIW

MSP-Greg replied:
I'm not seeing frozen string errors...

eregon replied:
Since this job applies only for jruby and truffleruby, how about simply removing the conditional and always `--enable=frozen-string-literal`?

nateberkopec commented:
@MSP-Greg In this PR, let's also go ahead and delete Travis.yml 🎉

MSP-Greg commented:
> delete Travis.yml

Done.  Working on getting JRuby to run ssl tests...

nateberkopec commented:
I'll probably open up a separate PR later to let JRuby skip those integration tests outside the minitest suite that it keeps failing on.

MSP-Greg commented:
I looked at getting ssl working on JRuby.  There are methods and constants that are defined in MRI Puma that aren't defined in JRuby Puma.  Below is what I found:

```ruby
Puma::MiniSSL::Engine#init?
Puma::MiniSSL::Engine#shutdown

Puma::MiniSSL::OPENSSL_VERSION
Puma::MiniSSL::OPENSSL_LIBRARY_VERSION
Puma::MiniSSL::OPENSSL_NO_SSL3
Puma::MiniSSL::OPENSSL_NO_TLS1
Puma::MiniSSL::OPENSSL_NO_TLS1_1
```

MSP-Greg commented:
JRuby - I've got code that allows running all the SSL tests on JRuby.

They pass, but I'd call them false positives.  Hence, maybe JRuby should be moved to the 'allow-failure' until it really passes the SSL tests?

Thoughts?

MSP-Greg commented:
> as it allows further regressions on JRuby

Good point.

MSP-Greg commented:
Rebased and made some changes.

1. Added Windows Ruby 2.2, needed minor change in extconf.rb

2. Made adjustments to JRuby so it would pass CI.  It is missing `Puma::MiniSSL::Engine#init?` and `Puma::MiniSSL::Engine#shutdown` methods, which may be affecting it's throughput.

3. I suspect that Actions may be implementing a better 'allow-failure' for jobs, and they're also changing matrix additions.  Hence, the separate job in the workflow may be able to be folded into the normal Ubuntu/macOS job in the future...

I see that the Ubuntu JRuby job failed.  Intermittent, as it passed in my fork.

nateberkopec commented:
Awesome work @MSP-Greg, thanks,
